File: Android/app/src/main/java/com/example/tutorial/CustomAdapter.java
package com.example.tutorial;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

public class CustomAdapter extends BaseAdapter {
    private Context context;
    private String[] data;

    public CustomAdapter(Context context, String[] data) {
        this.context = context;
        this.data = data;
    }


    @Override
    public int getCount() {
        return data.length;
    }

    @Override
    public Object getItem(int i) {
        return data[i];
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public View getView(int i, View ConvertView, ViewGroup parent) {

        ViewHolder viewHolder;
        if (ConvertView == null) {
            viewHolder = new ViewHolder();
            // Inflate the layout here
            ConvertView = LayoutInflater.from(context)
                    .inflate( R.layout.custom_list_items, parent ,false);
            // Initialize the views here
            viewHolder.textView = ConvertView.findViewById(R.id.textView);
            ConvertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) ConvertView.getTag();
        }

        // Set the data to the views here
        viewHolder.textView.setText(data[i]);

        return ConvertView;
    }

    static class ViewHolder {
        // Declare the views here
        TextView textView;
    }
}

File: kotlin/PlanetApp/app/src/main/java/com/example/planetapp/CustomAdapter.kt
package com.example.planetapp

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView

class CustomAdapter (context: Context, planets:List<Planets>) :BaseAdapter(){
    private val mContext:Context = context
    private val mPlanets:List<Planets> = planets


    override fun getCount(): Int {
        return mPlanets.size
    }

    override fun getItem(p0: Int): Any {

        return mPlanets[p0]
    }

    override fun getItemId(p0: Int): Long {
        return p0.toLong()
    }

    override fun getView(p0: Int, p1: View?, p2: ViewGroup?): View {

        val inflater = mContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val view:View
        if(p1 == null){
            view = inflater.inflate(R.layout.item_list_layout, null)
        }else{
            view = p1
        }

        val item = getItem(p0) as Planets

       val titleTxt =  view.findViewById<TextView>(R.id.planetName)
        val moonCountTxt = view.findViewById<TextView>(R.id.moonCount)
        val image = view.findViewById<ImageView>(R.id.planetImage)

        titleTxt.text = item.title
        moonCountTxt.text = item.moonCount
        image.setImageResource(item.image)
        return view



    }


}